### PR TITLE
Finer control of locate_tiles

### PR DESCRIPTION
**Pull request by kheffah**
Finer control of locate_tiles includin

1. Allowing user to pass tiless to avoid re-extraction
2. Optionally not fading image
3. Passing individual colors to specific tiles (eg to emphasize specific regions of interest).

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://github.com/histolab/histolab/blob/master/CONTRIBUTING.md) document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by alessiamarcolini** [comment]
Hey @kheffah thanks for this PR!! First of all, could you split this commit into 3 separate commits?

**Comment by kheffah** [comment]
@alessiamarcolini All done :)

File: src/histolab/tiler.py [hunk 1]
```diff
@@ -90,25 +92,39 @@ def locate_tiles(
             The alpha level to be applied to the rescaled slide, default to 128.
         outline: str
             The outline color for the tile annotations, default to 'red'.
+        linewidth: int
+            Thickness of line used to draw tiles
+        tiles: Iterable
+            Tiles to visualize. Optional, will be extracted if None.
```
Comment (ernestoarbitrio) [inline src/histolab/tiler.py:98]: I think this part of docstring is not clear. I would use something more verbose

File: src/histolab/tiler.py [hunk 2]
```diff
@@ -90,25 +92,39 @@ def locate_tiles(
...
 
         Returns
         -------
         PIL.Image.Image
             PIL Image of the rescaled slide with the extracted tiles outlined
         """
         img = slide.scaled_image(scale_factor)
-        img.putalpha(alpha)
+        if alpha is not None:
```
Comment (ernestoarbitrio) [inline src/histolab/tiler.py:106]: Why alpha can be none if it has a default value? I think that alpha cannot be none and if a user passes none an error would be expected, that is fine I think

**Review by ernestoarbitrio — CHANGES_REQUESTED** [review]
@kheffah thanks for this contribution. That's great. Just a couple of comments

File: src/histolab/tiler.py (hunk 2 above)
Comment (kheffah) [reply src/histolab/tiler.py:106]: @ernestoarbitrio Actually this is on purpose. The default value is to fade out the slide image. But some users (myself included) prefer _not_ to fade out the background image, in which case they'd pass ``alpha=None``.  This is similar to how PIL.MAX_IMAGE_PIXELS has a default numerical value, but users can opt out by passing it as ``None``. Let me know if that makes sense to you :)

Comment (ernestoarbitrio) [reply src/histolab/tiler.py:106]: Ok, I see, but IMHO `alpha` has to be 255 that means no transparency and so the image will be without any fade i think. This way we could get rid of that `if` statement.

File: src/histolab/tiler.py [hunk 3]
```diff
@@ -73,7 +73,9 @@ def locate_tiles(
         extraction_mask: BinaryMask = BiggestTissueBoxMask(),
         scale_factor: int = 32,
         alpha: int = 128,
-        outline: str = "red",
+        outline: Union[str, List] = "red",
+        linewidth: int = 1,
+        tiles: Iterable = None,
```
Comment (ernestoarbitrio) [inline src/histolab/tiler.py:78]: I'd like to see what kind of object for the iterable e.g. `Iterable[str]` this way is more meaningful when we read the code and we know that we're expecting an iterable of `Tile` objects or something like that. Hope I have been clear 🤷🏽‍♂️

File: src/histolab/tiler.py [hunk 4]
```diff
@@ -90,25 +92,39 @@ def locate_tiles(
...
-        )
-        tiles_coords = (tile[1] for tile in tiles)
-        for coords in tiles_coords:
+        if tiles is None:
+            tiles = (
+                self._tiles_generator(slide, extraction_mask)[0]
+                if isinstance(self, ScoreTiler)
+                else self._tiles_generator(slide, extraction_mask)
+            )
+        tiles_coords = [tile[1] for tile in tiles]
```
Comment (ernestoarbitrio) [inline src/histolab/tiler.py:116]: i would keep it as a `tuple` as before:
`tiles_coords = (tile[1] for tile in tiles)`

File: src/histolab/tiler.py [hunk 5]
```diff
@@ -90,25 +92,39 @@ def locate_tiles(
...
+
+        # maybe user specified colors for each tile
```
Comment (ernestoarbitrio) [inline src/histolab/tiler.py:118]: comments like this are not meaningful in my opinion. We should use comments for specify something that could be not clear in the future. So in this case the comment should be removed if the statement or the code portion is clear enough, otherwise the statement has to be improved XD

File: src/histolab/tiler.py [hunk 6]
```diff
@@ -73,7 +73,9 @@ def locate_tiles(
         extraction_mask: BinaryMask = BiggestTissueBoxMask(),
         scale_factor: int = 32,
         alpha: int = 128,
-        outline: str = "red",
+        outline: Union[str, List] = "red",
```
Comment (ernestoarbitrio) [inline src/histolab/tiler.py:76]: I'm not sure this will work properly. We are using a union of string and list for the sake of the retro-compatibility but this can be a problem:
1. This implies a lot of if statement in the code that in turn requires a bunch of unit/integration tests with some edge cases.
2. the outline when it's a list and the length of the list is greater or less than the number of the tiles we will throw several `IndexError` or wrong color according to the size of the lists. See line `124`.

I wonder if we have to simplify this method or in some way introduce a validator for the outline.

If something is not clear please let me know 😄

File: tests/unit/test_tiler.py [hunk 7]
```diff
@@ -777,6 +778,53 @@ def it_knows_whether_coordinates_are_within_extraction_mask(
...
         assert coords_within_extraction_mask == expected_result
 
+    @pytest.mark.parametrize(
+        "alpha, pass_tiles, outline, expected_topleft",
+        (
+            (128, False, "red", (255, 255, 155, 155)),
+            (None, True, [[0.1, 0.2, 0.3], [0.4, 0.5, 0.6]], (102, 102, 155, 155)),
+        ),
+    )
+    def it_can_locate_tiles(
```
Comment (ernestoarbitrio) [inline tests/unit/test_tiler.py:788]: I can buy this unit test, but given the several changes I'd like to see at least 1 integration test for the edge cases used here

**Review by ernestoarbitrio — CHANGES_REQUESTED** [review]
Per comments, please let me know if something is not clear enough

File: src/histolab/tiler.py (hunk 3 above)
Comment (kheffah) [reply src/histolab/tiler.py]: OK, all addressed.

File: src/histolab/tiler.py (hunk 4 above)
Comment (kheffah) [reply src/histolab/tiler.py]: I agree. Now everything is a generator to preserve memory and time.

File: src/histolab/tiler.py (hunk 6 above)
Comment (kheffah) [reply src/histolab/tiler.py]: That's a great point! I added a new generator that zips outline and tiles and throws errors for the edge cases. These edge cases were all added to unit tests.

File: tests/unit/test_tiler.py (hunk 7 above)
Comment (kheffah) [reply tests/unit/test_tiler.py:836]: Actually, the edge cases are best resolved with a unit test. I added a unit test specifically for the edge cases. I also added integration test cases to make sure the new outline functionality produces the expected output.

Comment (ernestoarbitrio) [reply tests/unit/test_tiler.py:836]: ok nice

File: src/histolab/tiler.py [hunk 8]
```diff
@@ -100,19 +122,62 @@ def locate_tiles(
...
+        if tiles is None:
+            tiles = (
+                self._tiles_generator(slide, extraction_mask)[0]
+                if isinstance(self, ScoreTiler)
+                else self._tiles_generator(slide, extraction_mask)
+            )
         tiles_coords = (tile[1] for tile in tiles)
-        for coords in tiles_coords:
+
+        for coords, one_outline in self._tile_coords_and_outline_generator(
```
Comment (ernestoarbitrio) [inline src/histolab/tiler.py:133]: I'm not sure `_tile_coords_and_outline_generator` is the right approach here. My general concern is that basically we are checking for the outline type and then we return the coord and the outline as they are passed to the method if they respect all the constraint. So it's just a validator, IMHO i do prefer to have a validator as you had in one of your pervious commit. Also i think that we could have loss of performance with a large number of Tiles and outlines 🤷🏽‍♂️

File: src/histolab/tiler.py [hunk 9]
```diff
@@ -100,19 +122,62 @@ def locate_tiles(
...
+        ):
             rescaled = scale_coordinates(coords, slide.dimensions, img.size)
-            draw.rectangle(tuple(rescaled), outline=outline)
+            draw.rectangle(tuple(rescaled), outline=one_outline, width=linewidth)
         return img
 
     # ------- implementation helpers -------
 
+    @staticmethod
+    def _tile_coords_and_outline_generator(
```
Comment (ernestoarbitrio) [inline src/histolab/tiler.py:143]: FYI: We decided that all methods and properties (public or private) should be A-Z ordered so please keep in mind that all new method should be placed in the right position :D

**Review by ernestoarbitrio — CHANGES_REQUESTED** [review]
I think we're close. Could you address my last comments please :D

File: tests/integration/test_tiler.py [hunk 10]
```diff
@@ -18,13 +18,32 @@
 
 class DescribeRandomTiler:
     @pytest.mark.parametrize(
-        "fixture_slide, binary_mask, tile_size, level, check_tissue, expectation",
+        "fixture_slide, binary_mask, tile_size, level, outline, check_tissue, expectation",
```
Comment (ernestoarbitrio) [inline tests/integration/test_tiler.py:21]: this line is 91 char. Be sure you have black formatter before commit

File: src/histolab/tiler.py (hunk 8 above)
Comment (kheffah) [reply src/histolab/tiler.py:133]: I beg to differ here. It's not just a validator .. depending on the type of ``outline`` argument, it either calls ``for coords in tiles_coords`` or ``for coords, one_outline in zip_longest(tiles_coords, outlines)``. So if the outline is a ``str`` it would return the same value for all coords, whereas if it is an ``Iterable`` it would fetch the value corresponding to the relevant tile and throw an exception is the lengths don't match. Even if I wanted to use a validator I couldn't because the number of tiles is not known beforehand in ``GridTiler`` so the only way to check the difference in length between it and the ``outline`` ``Iterable`` is using something like the ``zip_longest``.

File: src/histolab/tiler.py (hunk 9 above)
Comment (kheffah) [reply src/histolab/tiler.py:162]: @ernestoarbitrio OK, addressed.

File: tests/integration/test_tiler.py (hunk 10 above)
Comment (kheffah) [reply tests/integration/test_tiler.py:21]: @ernestoarbitrio Actually I did use black .. it automatically runs on all my commits with the precommit-hooks. I ran it again now and it didn't make any changes to that file. Are you sure 91 char is the maximum allowed?

Comment (ernestoarbitrio) [reply tests/integration/test_tiler.py:21]: 88 is the max

Comment (ernestoarbitrio) [reply tests/integration/test_tiler.py:21]:
```
.\tests\integration\test_tiler.py:21:89: E501 line too long (91 > 88 characters)
```

File: src/histolab/tiler.py (hunk 8 above)
Comment (ernestoarbitrio) [reply src/histolab/tiler.py:133]: __gotcha__ ok

File: tests/integration/test_tiler.py (hunk 10 above)
Comment (ernestoarbitrio) [reply tests/integration/test_tiler.py:21]: black doesn't take care of single strings > than 88 chars

Comment (kheffah) [reply tests/integration/test_tiler.py]: OK, fixed.

**Review by ernestoarbitrio — APPROVED** [review]

File: src/histolab/tiler.py [hunk 11]
```diff
@@ -84,12 +87,31 @@ def locate_tiles(
...
             BinaryMask object defining how to compute a binary mask from a Slide.
             Default `BiggestTissueBoxMask`
-        scale_factor: int
+        scale_factor: int, optional
             Scaling factor for the returned image. Default is 32.
-        alpha: int
-            The alpha level to be applied to the rescaled slide, default to 128.
-        outline: str
-            The outline color for the tile annotations, default to 'red'.
+        alpha: int, optional
```
Comment (alessiamarcolini) [inline src/histolab/tiler.py:92]:
```suggestion
        alpha : int, optional
```

File: src/histolab/tiler.py [hunk 12]
```diff
@@ -84,12 +87,31 @@ def locate_tiles(
...
+            The alpha level to be applied to the rescaled slide. Default is 128.
+        outline: Union[str, Iterable[str], Iterable[Tuple[int]]], optional
```
Comment (alessiamarcolini) [inline src/histolab/tiler.py:94]:
```suggestion
        outline : Union[str, Iterable[str], Iterable[Tuple[int]]], optional
```

File: src/histolab/tiler.py [hunk 13]
```diff
@@ -84,12 +87,31 @@ def locate_tiles(
...
+            each color is your assigned color for the corresponding individual
+            tile. This list can be a list of matplotlib-style string colors, or
+            it maybe a list of tuples of ints in the [0, 255] range, each of
+            length 3, representing the red, green and blue color for each tile.
+            For example, if you have two tiles that you want to be colored
+            yellow, you can pass this argument as any of the following ..
+            - 'yellow'
+            - ['yellow', 'yellow']
+            - [(255, 255, 0), (255, 255, 0)]
+        linewidth: int, optional
```
Comment (alessiamarcolini) [inline src/histolab/tiler.py:107]:
```suggestion
        linewidth : int, optional
```

File: src/histolab/tiler.py [hunk 14]
```diff
@@ -84,12 +87,31 @@ def locate_tiles(
...
+            Thickness of line used to draw tiles. Default is 1.
+        tiles: Iterable[Tile], optional
```
Comment (alessiamarcolini) [inline src/histolab/tiler.py:109]:
```suggestion
        tiles : Optional[Iterable[Tile]], optional
```

File: src/histolab/tiler.py [hunk 15]
```diff
@@ -73,7 +74,9 @@ def locate_tiles(
         extraction_mask: BinaryMask = BiggestTissueBoxMask(),
         scale_factor: int = 32,
         alpha: int = 128,
-        outline: str = "red",
+        outline: Union[str, Iterable[str], Iterable[Tuple[int]]] = "red",
+        linewidth: int = 1,
+        tiles: Iterable[Tile] = None,
```
Comment (alessiamarcolini) [inline src/histolab/tiler.py:79]:
```suggestion
        tiles: Optional[Iterable[Tile]] = None,
```

File: src/histolab/tiler.py [hunk 16]
```diff
@@ -84,12 +87,31 @@ def locate_tiles(
...
+            - 'yellow'
+            - ['yellow', 'yellow']
+            - [(255, 255, 0), (255, 255, 0)]
+        linewidth: int, optional
+            Thickness of line used to draw tiles. Default is 1.
+        tiles: Iterable[Tile], optional
+            Tiles to visualize. Will be extracted if None. Default is None.
+            You may decide to provide this argument if you do not want the
+            tiles to be re-extracted for visualization if you already have
+            the tiles in hand from, say, the ``_tiles_generator`` method of
```
Comment (alessiamarcolini) [inline src/histolab/tiler.py:113]: mmm why are we making a reference to a private method here..?

File: src/histolab/tiler.py [hunk 17]
```diff
@@ -132,6 +158,45 @@ def _has_valid_tile_size(self, slide: Slide) -> bool:
...
+    @staticmethod
+    def _tile_coords_and_outline_generator(
+        tiles_coords: Iterable[CoordinatePair],
+        outlines: Union[str, List[str], List[Tuple[int]]],
+    ) -> Union[str, Tuple[int]]:
+        """Zip tile coordinates and outlines from tile and outline iterators.
+
+        Parameters
+        ----------
+        tiles_coords: Iterable[CoordinatePair]
```
Comment (alessiamarcolini) [inline src/histolab/tiler.py:170]:
```suggestion
        tiles_coords : Iterable[CoordinatePair]
```

File: src/histolab/tiler.py [hunk 18]
```diff
@@ -132,6 +158,45 @@ def _has_valid_tile_size(self, slide: Slide) -> bool:
...
+            Coordinates referring to the upper left and lower right corners.
+        outlines: Union[str, Iterable[str], Iterable[Tuple[int]]]
```
Comment (alessiamarcolini) [inline src/histolab/tiler.py:172]:
```suggestion
        outlines : Union[str, Iterable[str], Iterable[Tuple[int]]]
```

File: src/histolab/tiler.py [hunk 19]
```diff
@@ -132,6 +158,45 @@ def _has_valid_tile_size(self, slide: Slide) -> bool:
...
+    def _tile_coords_and_outline_generator(
+        tiles_coords: Iterable[CoordinatePair],
+        outlines: Union[str, List[str], List[Tuple[int]]],
+    ) -> Union[str, Tuple[int]]:
+        """Zip tile coordinates and outlines from tile and outline iterators.
+
+        Parameters
+        ----------
+        tiles_coords: Iterable[CoordinatePair]
+            Coordinates referring to the upper left and lower right corners.
```
Comment (alessiamarcolini) [inline src/histolab/tiler.py:171]:
```suggestion
            Coordinates referring to the tiles' upper left and lower right corners.
```

File: src/histolab/tiler.py [hunk 20]
```diff
@@ -132,6 +158,45 @@ def _has_valid_tile_size(self, slide: Slide) -> bool:
...
+        outlines: Union[str, Iterable[str], Iterable[Tuple[int]]]
+            See docstring for ``locate_tiles`` for details.
+
+        Yields
+        -------
+        CoordinatePair
+            Coordinates referring to the upper left and lower right corners.
```
Comment (alessiamarcolini) [inline src/histolab/tiler.py:178]:
```suggestion
            Coordinates referring to the tiles' upper left and lower right corners.
```

File: src/histolab/tiler.py [hunk 21]
```diff
@@ -132,6 +158,45 @@ def _has_valid_tile_size(self, slide: Slide) -> bool:
...
+        Union[str, Tuple[int]]
+            Fixed outline depending on user input to used by method ``locate_tiles``.
+        """
+        if isinstance(outlines, str):
+            for coords in tiles_coords:
+                yield coords, outlines
+
+        elif hasattr(outlines, "__iter__"):
+            for coords, one_outline in zip_longest(tiles_coords, outlines):
+                if any(j is None for j in (coords, one_outline)):
```
Comment (alessiamarcolini) [inline src/histolab/tiler.py:188]: can this be replaced with
`if None in (coords, one_outline)` ?

File: tests/unit/test_tiler.py [hunk 22]
```diff
@@ -777,6 +778,98 @@ def it_knows_whether_coordinates_are_within_extraction_mask(
...
+                "str, Iterable[str], or Iterable[List[int]]",
+            ),
+        ),
+    )
+    def it_throws_error_with_invalid_tile_outline(
+        self, request, tmpdir, outline, n_tiles, error_msg
+    ):
+        slide, _ = base_test_slide(tmpdir, PILIMG.RGBA_COLOR_500X500_155_249_240)
+        _extract_tile = method_mock(request, Slide, "extract_tile")
+
```
Comment (alessiamarcolini) [inline tests/unit/test_tiler.py:807]:
```suggestion
```

File: tests/unit/test_tiler.py [hunk 23]
```diff
@@ -777,6 +778,98 @@ def it_knows_whether_coordinates_are_within_extraction_mask(
...
+        coords = CP(0, 10, 0, 10)
+        tile = Tile(PILIMG.RGBA_COLOR_500X500_155_249_240, coords)
+        mock_tiles = [(tile, coords)] * n_tiles
+
```
Comment (alessiamarcolini) [inline tests/unit/test_tiler.py:811]:
```suggestion
```

File: tests/unit/test_tiler.py [hunk 24]
```diff
@@ -777,6 +778,98 @@ def it_knows_whether_coordinates_are_within_extraction_mask(
...
+        _tiles_generator = method_mock(request, GridTiler, "_tiles_generator")
+        _tiles_generator.return_value = mock_tiles
+
```
Comment (alessiamarcolini) [inline tests/unit/test_tiler.py:814]:
```suggestion
```

File: tests/unit/test_tiler.py [hunk 25]
```diff
@@ -777,6 +778,98 @@ def it_knows_whether_coordinates_are_within_extraction_mask(
...
+        _has_enough_tissue = method_mock(request, Tile, "has_enough_tissue")
+        _has_enough_tissue.side_effect = True
+        _grid_coordinates_generator = method_mock(
+            request, GridTiler, "_grid_coordinates_generator"
+        )
+        _grid_coordinates_generator.return_value = [CP(0, 10, 0, 10), CP(0, 10, 0, 10)]
+        _extract_tile.side_effect = mock_tiles
+        grid_tiler = GridTiler((10, 10))
```
Comment (alessiamarcolini) [inline tests/unit/test_tiler.py:822]:
```suggestion
        grid_tiler = GridTiler((10, 10))
        
```

File: tests/unit/test_tiler.py [hunk 26]
```diff
@@ -777,6 +778,98 @@ def it_knows_whether_coordinates_are_within_extraction_mask(
...
+                "The parameter ``outline`` should be of type: "
+                "str, Iterable[str], or Iterable[List[int]]",
+            ),
+        ),
+    )
+    def it_throws_error_with_invalid_tile_outline(
+        self, request, tmpdir, outline, n_tiles, error_msg
+    ):
+        slide, _ = base_test_slide(tmpdir, PILIMG.RGBA_COLOR_500X500_155_249_240)
+        _extract_tile = method_mock(request, Slide, "extract_tile")
```
Comment (alessiamarcolini) [inline tests/unit/test_tiler.py:806]: I'm not sure I understand why we need this

File: tests/unit/test_tiler.py [hunk 27]
```diff
@@ -777,6 +778,98 @@ def it_knows_whether_coordinates_are_within_extraction_mask(
...
+        _tiles_generator = method_mock(request, GridTiler, "_tiles_generator")
+        _tiles_generator.return_value = mock_tiles
+
+        _has_enough_tissue = method_mock(request, Tile, "has_enough_tissue")
+        _has_enough_tissue.side_effect = True
+        _grid_coordinates_generator = method_mock(
+            request, GridTiler, "_grid_coordinates_generator"
+        )
+        _grid_coordinates_generator.return_value = [CP(0, 10, 0, 10), CP(0, 10, 0, 10)]
+        _extract_tile.side_effect = mock_tiles
```
Comment (alessiamarcolini) [inline tests/unit/test_tiler.py:821]: same

File: tests/unit/test_tiler.py [hunk 28]
```diff
@@ -777,6 +778,98 @@ def it_knows_whether_coordinates_are_within_extraction_mask(
...
+        binary_mask = BiggestTissueBoxMask()
+
+        img = grid_tiler.locate_tiles(
+            slide=slide,
+            extraction_mask=binary_mask,
+            alpha=alpha,
+            outline=outline,
+            tiles=None
+            if not pass_tiles
+            else list(grid_tiler._tiles_generator(slide, binary_mask)),
```
Comment (alessiamarcolini) [inline tests/unit/test_tiler.py:866]: if statements in test code are considered a bad practice, could you refactor this test to avoid this?
I think we already discussed this in a previous PR (now I don't remember...)

File: tests/unit/test_tiler.py [hunk 29]
```diff
@@ -777,6 +778,98 @@ def it_knows_whether_coordinates_are_within_extraction_mask(
...
+        )
+        img = np.uint8(img)
+
+        assert img.shape == (15, 15, 4 if alpha else 3)
```
Comment (alessiamarcolini) [inline tests/unit/test_tiler.py:870]: in this case you can parametrize this to avoid the if statement

**Review by alessiamarcolini — CHANGES_REQUESTED** [review]
some stylistic changes + unit test code shouldn't have conditionals (see comments)

File: tests/unit/test_tiler.py (hunk 29 above)
Comment (ernestoarbitrio) [reply tests/unit/test_tiler.py]: alpha doesn't mean anything now cause it cannot be None anymore ... so i remove it

File: tests/unit/test_tiler.py (hunk 28 above)
Comment (ernestoarbitrio) [reply tests/unit/test_tiler.py]: I think we can use this if to parametrize a test without duplicating all the stuff... If statements are bad in case of assertions but i think we can buy this

File: tests/unit/test_tiler.py [hunk 30]
```diff
@@ -778,6 +779,73 @@ def it_knows_whether_coordinates_are_within_extraction_mask(
...
+        coords = CP(0, 10, 0, 10)
+        tile = Tile(PILIMG.RGBA_COLOR_500X500_155_249_240, coords)
+        mock_tiles = [(tile, coords), (tile, coords)]
+        method_mock(request, GridTiler, "_tiles_generator", return_value=mock_tiles)
+        method_mock(
+            request,
+            GridTiler,
+            "_grid_coordinates_generator",
+            return_value=[CP(0, 10, 0, 10), CP(0, 10, 0, 10)],
+        )
```
Comment (alessiamarcolini) [inline tests/unit/test_tiler.py:840]: why do we need this? this method is not called in the `locate_tiles` method, and we mock `_tiles_generator` that would call it

moreover if we mock something it's also better to assert that the mock has been called

**Review by alessiamarcolini — CHANGES_REQUESTED** [review]

File: tests/unit/test_tiler.py (hunk 30 above)
Comment (ernestoarbitrio) [reply tests/unit/test_tiler.py:840]: > why do we need this? this method is not called in the `locate_tiles` method, and we mock `_tiles_generator` that would call it
yes u right this is not needed
> 
> moreover if we mock something it's also better to assert that the mock has been called
that's not true Alessia, if we mock someting for the sake of isolation it's not mandatory to assert that something has been called ... we mock a lot of property for example and we do not assert anything 🤷🏽‍♂️

**Review by ernestoarbitrio — APPROVED** [review]
